src/util/result.h

```diff
@@ -7,42 +7,119 @@
+    Result() = default;
+
+    //! Error case constructor for a single error.
+    Result(Error, bilingual_str error)
+    {
+        m_errors.emplace_back(std::move(error));
     }
-    T ReleaseObj()
+
+    //! Error case constructor for a chained error.
+    Result(ErrorChain, bilingual_str error, Result<void>&& previous) : m_errors{std::move(previous.m_errors)}, m_warnings{std::move(previous.m_warnings)}
     {
-        assert(HasRes());
-        return std::move(std::get<T>(m_variant));
+        m_errors.emplace_back(std::move(error));
     }
 
-    /* In case of failure, the error cause */
-    const bilingual_str& GetError() const {
-        assert(!HasRes());
-        return std::get<bilingual_str>(m_variant);
+    //! Success check.
+    operator bool() const { return m_errors.empty(); }
+
+    //! Error retrieval.
+    const std::vector<bilingual_str>& GetErrors() const { return m_errors; }
+    std::tuple<const std::vector<bilingual_str>&, const std::vector<bilingual_str>&> GetErrorsAndWarnings() const { return {m_errors, m_warnings}; }
+};
+
+template<typename T>
+class Result : public Result<void>
+{
+protected:
+    T m_result;
+
+public:
+    //! Constructors that forward to the base class and pass additional arguments to m_result.
+    template<typename... Args>
+    Result(Args&&... args) : m_result{std::forward<Args>(args)...} {}
+    template<typename Str, typename...Args>
+    Result(Error, Str&& str, Args&&... args) : Result<void>{Error{}, std::forward<Str>(str)}, m_result{std::forward<Args>(args)...} {};
+    template<typename Str, typename Prev, typename...Args>
+    Result(ErrorChain, Str&& str, Prev&& prev, Args&&... args) : Result<void>{ErrorChain{}, std::forward<Str>(str), std::forward<Prev>(prev)}, m_result{std::forward<Args>(args)...} {};
+
+    //! std::optional methods, so Result<T> can be easily swapped for
+    //! std::optional<T> to add error reporting to existing code or remove it if
+    //! it is no longer needed.
+    bool has_value() const { return m_errors.empty(); }
+    const T& value() const { return m_result; }
+    T& value() { return m_result; }
+    template<typename U> T value_or(const U& default_value) const
+    {
+        return has_value() ? value() : default_value;
     }
+    const T* operator->() const { return &value(); }
+    const T& operator*() const { return value(); }
+    T* operator->() { return &value(); }
+    T& operator*() { return value(); }
+};
+
+//! Helper method to retrieve a simple error string from Result<T> or
+//! Result<void>.
+bilingual_str ErrorDescription(const Result<void>& result);
+} // namespace util
 
+/**
+ * Backwards-compatible interface for util::Result class. New code should prefer
+ * util::Result class which supports returning error information along with
+ * result information and supports returing `void` and `bilingual_str` results.
+*/
+template<class T>
+class BResult {
+private:
+    util::Result<std::optional<T>> m_result;
```

Comment:
I don't think this is too helpful to force most call sites (that can use BResult or use it today) into a double wrapping `util::Result<std::optional<T>>`. I liked a single wrapping better.

Reply:
> I don't think this is too helpful to force most call sites (that can use BResult or use it today) into a double wrapping `util::Result<std::optional<T>>`. I liked a single wrapping better.

Right this was supposed to be only temporary within the PR. A followup commit basically like 6a06a7c3ad264c0ddf4904949839be1ade8011a5 (from #25665) would replace `BResult<T>` with `util::Result<T>`